Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,448 @@
+  let simpNames := simpOnlyBuiltins ++ [
+    ``add_comm, ``add_assoc, ``add_left_comm,  -- for `HAdd.hAdd`
+    ``mul_comm, ``mul_assoc, ``mul_left_comm,  -- for `HMul.hMul`
+    ``and_comm, ``and_assoc, ``and_left_comm,  -- for `and`
+    ``or_comm,  ``or_assoc,  ``or_left_comm,   -- for `or`
+    ``max_comm, ``max_assoc, ``max_left_comm,  -- for `max`
+    ``min_comm, ``min_assoc, ``min_left_comm   -- for `min`
+    ]
+  let simpThms := ← simpNames.foldlM (·.addConst ·) ({} : SimpTheorems)
+  return { simpTheorems := #[simpThms] }
+
+/-- `reorderAndSimp mv op instr` takes as input an `MVarId`  `mv`, the name `op` of a binary
+operation and a list of "instructions" `instr` that it passes to `permuteExpr`.
+
+* It creates a version `permuted_mv` of `mv` with subexpressions representing `op`-sums reordered
+  following `instructions`.
+* It produces 2 temporary goals by applying `Eq.mpr` and unifying the resulting meta-variable with
+  `permuted_mv`: `[⊢ mv = permuted_mv, ⊢ permuted_mv]`.
+* It tries to solve the goal `mv = permuted_mv` by a simple-minded `simp` call, using the
+  `op`-analogues of `add_comm, add_assoc, add_left_comm`.
+-/
+def reorderAndSimp (mv : MVarId) (op : Name) (instr : List (Expr × Bool)) :
+    MetaM (List MVarId) := do
+  let permExpr := ← permuteExpr op (← mv.getType'') instr
+  -- generate the implication `permutedMv → mv = permutedMv → mv`
+  let eqmpr := ← mkAppM ``Eq.mpr #[← mkFreshExprMVar (← mkEq (← mv.getType) permExpr)]
+  let twoGoals := ← mv.apply eqmpr
+  guard (twoGoals.length == 2) <|>
+    throwError m!"There should only be 2 goals, instead of {twoGoals.length}"
+  -- `permGoal` is the single goal `mv_permuted`, possibly more operations will be permuted later on
+  let permGoal := ← twoGoals.filterM fun v => return !(← v.isAssigned)
+  match ← (simpGoal (permGoal[1]!) (← move_oper_simpCtx)) with
+    | (some x, _) => throwError m!"'move_oper' could not solve {indentD x.2}"
+    | (none, _) => return permGoal
+
+section parsing
+open Elab Parser Tactic
+
+/-- `parsing` parses an input of the form `[a, ← b, _ * (1 : ℤ)]`, consisting of a list of
+terms, each optionally preceded by the arrow `←`.
+It unifies the terms with the atoms for the operation `op` in the expression `tgt`, returning
+* the lists of pairs of a matched subexpression with `true` for an arrow `←` and `false` otherwise;
+* an array of error messages;
+* an array of debugging messages.
+
+E.g. convert `[a, ← b, _ * (1 : ℤ)]` to `pairs = ([(a, false), (z * 1, false)], , [(b, true)]`, if
+`b` does not match a subexpression of `tgt` and the first summand involving a multiplication by `1`
+is `z * 1`.
+-/
+def parsing (rws : TSyntax `Lean.Parser.Tactic.rwRuleSeq) (op : Name) (tgt : Expr) :
+    TermElabM (List (Expr × Bool) × (List MessageData × List Syntax) × Array MessageData) :=
+do match rws with
+  | `(rwRuleSeq| [$rs,*]) => do
+    let pairs := ← rs.getElems.mapM fun rstx => do
+      let r : Syntax := rstx
```

Comment:
I don't get the `let` line: why not use `rstx` everywhere? Or is there something I missed here?

Reply:
I see! The reason is `TSyntax` is a wrapper around `Syntax` and the `let` is somewhat hiding a cast. I guessed so, and I don't know a better way to express it, so this way is fine.